src/Upload.tsx

```diff
@@ -0,0 +1,151 @@
+import { KalturaClient } from "kaltura-typescript-client";
+import { KalturaUploadToken } from "kaltura-typescript-client/api/types";
+import { createUploadToken, uploadFile } from "KalturaModule";
+import * as React from "react";
+import { useEffect, useReducer } from "react";
+import { useErrorHandler } from "react-error-boundary";
+import { reducer } from "UploadReducer";
+
+/**
+ * Generate a compressed representation of the number of provided bytes.
+ *
+ * Taken from https://gist.github.com/lanqy/5193417#gistcomment-3590752
+ *
+ * @param bytes The number to convert to a SI unit
+ */
+const bytesToSize = (bytes: number): string => {
+  if (bytes === 0) {
+    return "";
+  }
+
+  const sizes: string[] = ["Bytes", "KB", "MB", "GB", "TB"];
+  const i: number = parseInt(
+    Math.floor(Math.log(bytes) / Math.log(1024)).toString()
+  );
+  if (i === 0) {
+    return `${bytes} ${sizes[i]}`;
+  }
+  return `${(bytes / Math.pow(1024, i)).toFixed(1)} ${sizes[i]}`;
+};
+
+export interface UploadProps {
+  /**
+   * A list of file extensions which should be allowed.
+   */
+  allowedTypes: string[];
+  /**
+   * A prefix to add to the DOM `id` for the root of this component.
+   */
+  idPrefix: string;
+  /**
+   * A `KaltruaClient` that has been setup with valid session token and partner ID, possibly using
+   * `createClient` from `KalturaModule`.
+   */
+  kClient: KalturaClient;
+  /**
+   * Callback to be triggered upon a successful file upload.
+   *
+   * @param uploadResult details returned from the server about the upload attempt.
+   */
+  onUploadSuccessful: (uploadResult: KalturaUploadToken) => void;
+}
+
+/**
+ * A component providing a file selector and then integration to upload to a Kaltura instance
+ * configured by the provided `KalturaClient`.
+ */
+export const Upload = ({
+  allowedTypes,
+  idPrefix,
+  kClient,
+  onUploadSuccessful,
+}: UploadProps): JSX.Element => {
+  const [state, dispatch] = useReducer(reducer, { id: "start" });
+
+  useErrorHandler(state.id === "error" ? state.error : undefined);
+
+  useEffect(() => {
+    const dispatchError = (c: string | Error) =>
+      dispatch({ id: "failed", cause: c instanceof Error ? c : new Error(c) });
+
+    const doUpload = async (file: File): Promise<void> => {
+      try {
+        const uploadToken = await createUploadToken(kClient);
+        if (uploadToken) {
+          const uploadResult = await uploadFile(kClient, uploadToken, file);
+          uploadResult
+            ? dispatch({ id: "upload_success", uploadResult })
+            : dispatchError("Attempt to upload to Kaltura failed.");
+        } else {
+          dispatchError(
+            "Attempt to create Kaltura Upload Token failed, please try again."
+          );
+        }
+      } catch (e) {
+        if (e instanceof Error) {
+          dispatchError(e);
+        } else {
+          console.error(e);
```

Comment:
Interested to see what other types `e` could be.

Reply:
Heh, me too. :grin: 

Just a safety net in case we ever need to find out. But Typescript typing kind of forces this.